pkg/scraper/client/resource/decode.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"time"
+
+	"github.com/prometheus/common/model"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/metrics-server/pkg/storage"
+)
+
+const (
+	nameSpaceMetricName         = "namespace"
+	podNameMetricName           = "pod"
+	containerNameMetricName     = "container"
+	nodeCpuUsageMetricName      = model.LabelValue("node_cpu_usage_seconds_total")
+	nodeMemUsageMetricName      = model.LabelValue("node_memory_working_set_bytes")
+	containerCpuUsageMetricName = model.LabelValue("container_cpu_usage_seconds_total")
+	containerMemUsageMetricName = model.LabelValue("container_memory_working_set_bytes")
+)
+
+func decodeBatch(samples []*model.Sample, nodeName string) *storage.MetricsBatch {
+	res := &storage.MetricsBatch{
+		Nodes: make(map[string]storage.NodeMetricsPoint),
+		Pods:  make(map[apitypes.NamespacedName]storage.PodMetricsPoint),
+	}
+	node := storage.NodeMetricsPoint{}
+	pods := make(map[apitypes.NamespacedName]map[string]*storage.ContainerMetricsPoint)
+	for _, sample := range samples {
+		//get metrics from sample
+		switch sample.Metric[model.MetricNameLabel] {
+		case nodeCpuUsageMetricName, nodeMemUsageMetricName:
+			getNodeMetricsUsed(sample, &node)
+		case containerCpuUsageMetricName, containerMemUsageMetricName:
+			getContainerMetricsUsed(sample, pods)
+		}
+	}
+
+	if node.Timestamp.IsZero() || nodeName == "" {
+		klog.V(1).InfoS("Failed getting node metric's Timestamp or nodeName", "node", nodeName, "metric", node)
+	} else {
+		node.Name = nodeName
+		res.Nodes[nodeName] = node
+	}
+
+	for podRef, podMetric := range pods {
+		if len(podMetric) != 0 {
+			//drop container metrics when Timestamp is zero
```

Comment:
Would be check if we maintain the same behavior vs Summary implementation. One way to do this is separate PRs for updating metricsBatch/tests vs implementing resource.

Reply:
Yes, I would also be in favor of splitting the work into separate PRs to verify that.

To come back to this exact corner case, I wonder if it's really meaningful to handle the case where metrics have a timestamp equal to zero. As far as I am aware, this shouldn't happen with the Prometheus client.

That said, it made me think of another corner case. What would happen if kubelet isn't able to get memory or CPU metrics for a particular container of a pod. I saw multiple occurrences of this happening in OpenShift. In this particular case, the usages reported by metrics-server will be wrong since some containers were missed.
In my opinion, to prevent reporting incorrect data, we should check that we have metrics for all the running containers of a pod before returning its PodMetric. I would go as far as dropping the PodMetrics entirely if the data is wrong as it would explicitly show that metrics-server is unable to get the correct usages of this pod.

Reply:
@serathius @dgrisonnet , What should we do now?
do we need to  split the work into separate PRs to verify that?

Reply:
Yes, please apply the changes you plan to do to MetricsBatch in the first PR. In the second PR would be good to better show what happends in decoder tests, now you remove old file and add new. This makes it hard to review and compare. Could you move the decode.go test file instead of recreating it?

Sorry if this request seems strange, but this is a big change and there is a need to build a trust when making major rewrites. One way to do it that we proposed here is keeping buisiness logic changes from test changes separate. I hope that will this change it will be trivial to see what changes and what not. 

Please let me know if you need help. Making git surgery like this is not always easy so I would be happy to assist.

Reply:
ok.Let me   do it.